parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,191 @@
+
+
+class Scheduler(Enum):
+    Unknown = 0
+    Slurm = 1
+    PBS = 2
+    Cobalt = 3
+
+
+def get_slurm_hosts_list() -> List[str]:
+    cmd = "scontrol show hostname $SLURM_NODELIST"
+    b_output = subprocess.check_output(
+        cmd, stderr=subprocess.STDOUT, shell=True
+    )  # bytes
+    output = b_output.decode().strip().split()
+    return output
+
+
+def get_pbs_hosts_list() -> List[str]:
+    nodefile_name = os.environ["PBS_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_cobalt_hosts_list() -> List[str]:
+    nodefile_name = os.environ["COBALT_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_nodes_in_batchjob(scheduler: Scheduler) -> List[str]:
+    nodelist = []
+    if scheduler == Scheduler.Slurm:
+        nodelist = get_slurm_hosts_list()
+    elif scheduler == Scheduler.PBS:
+        nodelist = get_pbs_hosts_list()
+    elif scheduler == Scheduler.Cobalt:
+        nodelist = get_cobalt_hosts_list()
+    return nodelist
+
+
+def identify_scheduler() -> Scheduler:
+    if os.environ.get("SLURM_NODELIST"):
+        return Scheduler.Slurm
+    elif os.environ.get("PBS_NODEFILE"):
+        return Scheduler.PBS
+    elif os.environ.get("COBALT_NODEFILE"):
+        return Scheduler.Cobalt
+    else:
+        return Scheduler.Unknown
+
+
+class MPIResourceUnavailable(Exception):
+    """Raised if there are no free resources available for an MPI request"""
+
+    def __init__(self, requested: int, available: int):
+        self.requested = requested
+        self.available = available
+
+    def __str__(self):
+        return f"MPIResourceUnavailable(requested={self.requested} available={self.available}"
+
+
+class NoopScheduler:
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        self.pending_task_q = pending_task_q
+        self.pending_result_q = pending_result_q
+
+    def put_task(self, task) -> None:
+        return self.pending_task_q.put(task)
+
+    def get_result(self, block: bool, timeout: float):
+        return self.pending_result_q.get(block, timeout=timeout)
+
+    def put_result(self, result) -> None:
+        return self.pending_result_q.put(result)
+
+
+class MPITaskScheduler(NoopScheduler):
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        super().__init__(pending_task_q, pending_result_q)
+        self.scheduler = identify_scheduler()
+        # PriorityQueue is threadsafe
+        self._backlog_queue: queue.PriorityQueue = queue.PriorityQueue()
+        self._map_tasks_to_nodes: Dict[str, List[str]] = {}
+        self.available_nodes = get_nodes_in_batchjob(self.scheduler)
+        self._free_node_counter = SpawnContext.Value("i", len(self.available_nodes))
```

Comment:
or even a lock and a regular `List`, which would support `len`!

Reply:
I've converted the `nodes_q` to a regular queue since we expect only two different threads to access it. I still think it is nicer to use that `mp.Value` as a thread-safe counter since it comes with its own lock.